### ISSUE TITLE
macOS/Darwin does not implement pthread barriers

### ISSUE THREAD

**Issue by zrrVAQb0ae** [szpajder/dumpvdl2#6]
Not sure how high priority this is, but dumpvdl2 doesn't compile without modification since macOS Darwin doesn't implement POSIX thread barriers. There seem to be projects out there that have implemented this functionality easily themselves to avoid this compatibility issue, and thought it might be great if new users could just compile without problems.

This project for instance implemented pthread barriers to fix the compatibility issue:
https://github.com/pmwkaa/ioarena/issues/10

Another example implementation:
https://github.com/ademakov/DarwinPthreadBarrier

References to pthread barriers:

dumpvdl2.h:273:8: error: unknown type name 'pthread_barrier_t';
extern pthread_barrier_t demods_ready, samples_ready;

dumpvdl2.c:41:1: error: unknown type name 'pthread_barrier_t';
pthread_barrier_t demods_ready, samples_ready;

dumpvdl2.c:74:13: warning: implicit declaration of function 'pthread_arrier_init' is invalid in C99 [-Wimplicit-function-declaration]
	if(	(ret = pthread_barrier_init(&demods_ready, NULL, ctx->num_channels+1)) !=0 ||

dumpvdl2.c:573:3: warning: implicit declaration of function 'pthread_barrier_wait' is invalid in C99 [-Wimplicit-function-declaration]
		pthread_barrier_wait(&demods_ready);

Also for users trying to compile on macOS:

To fix reference to endian.h in avlc.h:22, change:

#include <endian.h>

to 

#include <machine/endian.h>

**Comment by szpajder** [comment]
I've added a feature branch here: https://github.com/szpajder/dumpvdl2/tree/macosx_port

Can you please test it? I don't have a Mac at hand. TIA.

**Comment by zrrVAQb0ae** [comment]
**_Tested the port, but sadly seems to have created some new issues. The errors seem to be centered around macOS' machine/endian.h not satisfying the endianness test in esis.h, and errors in esis.c related to the unknown member named "type" in esis_hdr_t. There are also various warnings, some which seem rather critical._**
  
  
<br/>
<br/>


**Errors:**
  
  
  
_In file included from esis.c:23:
./esis.h:44:2: error: Unsupported endianness
#error Unsupported endianness_
  
   
...
  
  
_esis.c:122:28: error: no member named 'type' in 'esis_hdr_t'
                hdr->pid, hdr->len, hdr->type, pdu->holdtime);
                                    ~~~  ^_ 
  
  

...
  
  
  
_esis.c:134:14: error: no member named 'type' in 'esis_hdr_t'
        switch(hdr->type) {
               ~~~  ^_
<br/>

_esis.c:154:60: error: no member named 'type' in 'esis_hdr_t'
        char *pdu_name = (char *)dict_search(esis_pdu_types, hdr->type);
                                                             ~~~  ^_
<br/>

_esis.c:158:14: error: no member named 'type' in 'esis_hdr_t'
        switch(hdr->type) {
               ~~~  ^_

  
  

  
<br/>
<br/>
<br/>




**Warnings:**
  
<br/>
<br/>  
  
Some of these seem less critical, but others like the math function warnings don't.
  
  
e.g.

<br/>
<br/>

*In file included from adsc.c:26:
In file included from /opt/local/include/glib-2.0/glib.h:30:
In file included from /opt/local/include/glib-2.0/glib/galloca.h:32:
/opt/local/include/glib-2.0/glib/gtypes.h:35:10: warning: non-portable path to file '<Time.h>'; specified path differs in case from file name on disk
      [-Wnonportable-include-path]
#include <time.h>
         ^~~~~~~~
         <Time.h>*




<br/>
<br/>


...


<br/>
<br/>



_/opt/local/include/glib-2.0/glib/gdate.h:273:51: warning: declaration of 'struct tm' will not be visible outside of this function [-Wvisibility]
                                           struct tm   *tm);
                                                  ^_
<br/>

_/opt/local/include/glib-2.0/glib/gdate.h:273:51: warning: declaration of /opt/local/include/glib-2.0/glib/gdate.h'struct tm': 273will: 51not:  be visiblewarning : outside ofdeclaration  thisof  function'struct tm'  [-Wvisibility]will 
not be visible outside of this function [-Wvisibility]
                                           struct tm   *tm);                                           struct tm   *tm);_

                                                  ^
                                                  ^

  
<br/> 
<br/>
 
  
...
  
  
<br/>  
<br/>


_avlc.c:158:97: warning: shifting a negative signed value is undefined [-Wshift-negative-value]
        return reverse((buf[0] >> 1) | (buf[1] << 6) | (buf[2] << 13) | ((buf[3] & 0xfe) << 20), 28) & ONES(28);
                                                                                                       ^~~~~~~~_

<br/>
<br/>

_./dumpvdl2.h:140:22: note: expanded from macro 'ONES'
#define ONES(x) ~(~0 << (x))
                  ~~ ^_

<br/>
<br/>

_avlc.c:304:2: warning: implicit declaration of function 'strftime' is invalid in C99 [-Wimplicit-function-declaration]
        strftime(ftime, sizeof(ftime), "%F %T %Z",
        ^_

<br/>
<br/>

_avlc.c:305:10: warning: implicit declaration of function 'gmtime' is invalid in C99 [-Wimplicit-function-declaration]
                (utc ? gmtime(&v->burst_timestamp.tv_sec) : localtime(&v->burst_timestamp.tv_sec)));_

                       ^

<br/>
<br/>

_avlc.c:305:47: warning: implicit declaration of function 'localtime' is invalid in C99 [-Wimplicit-function-declaration]
                (utc ? gmtime(&v->burst_timestamp.tv_sec) : localtime(&v->burst_timestamp.tv_sec)));_

<br/>
<br/>


_warning: implicit declaration of function 'sincosf' is invalid in C99 [-Wimplicit-function-declaration]
        sincosf(M_PI/(2 * npoles) + (p-1) * M_PI / npoles, &ip, &rp);
        ^_

<br/>
<br/>

_demod.c:366:3: warning: implicit declaration of function 'sincosf' is invalid in C99 [-Wimplicit-function-declaration]
                sincosf(2.0f * M_PI * (float)i / 256.0f, sin_lut + i, cos_lut + i);
                ^_

<br/>
<br/>

_decode.c:187:13: warning: shifting a negative signed value is undefined [-Wshift-negative-value]
                header &= ONES(TRLEN+HDRFECLEN);
                          ^~~~~~~~~~~~~~~~~~~~~_

<br/>
<br/>


_./dumpvdl2.h:140:22: note: expanded from macro 'ONES'
#define ONES(x) ~(~0 << (x))
                  ~~ ^_

<br/>
<br/>

_decode.c:193:16: warning: shifting a negative signed value is undefined [-Wshift-negative-value]
                if((header & ONES(TRLEN+HDRFECLEN)) != header) {
                             ^~~~~~~~~~~~~~~~~~~~~_

<br/>
<br/>

_./dumpvdl2.h:140:22: note: expanded from macro 'ONES'
#define ONES(x) ~(~0 << (x))
                  ~~ ^_

<br/>
<br/>



_decode.c:200:33: warning: shifting a negative signed value is undefined [-Wshift-negative-value]
                v->datalen = reverse(header & ONES(TRLEN), TRLEN);
                                              ^~~~~~~~~~~_

<br/>
<br/>


_./dumpvdl2.h:140:22: note: expanded from macro 'ONES'
#define ONES(x) ~(~0 << (x))
                  ~~ ^_



<br/>
<br/>
<br/>
		
		

**Sorry if this output isn't that helpful. I'll try to look at it in more detail when I get a chance, but hopefully this will help a bit.**
		
<br/>
<br/>
<br/>		

**_Full build log here: https://gist.github.com/zrrVAQb0ae/2abc8986fc129695eab99ebbb59dcf75_**

**Comment by szpajder** [comment]
Merged into stable and released in 1.6.0.

Closing.
